File: tests/Aspire.Hosting.CodeGeneration.Elixir.ExTests/test/support/fake_host.ex
defmodule Aspire.FakeHost do
  @moduledoc """
  A fake AppHost server for the transport tests.

  The fake host listens on a temporary Unix domain socket, accepts one
  connection, and parses LSP framed JSON-RPC messages. Each parsed request goes
  to the owner process as `{:fake_host, :request, message}`. An optional
  `:handler` function scripts the response.
  """

  use GenServer

  @type response ::
          {:result, term()}
          | {:error, term()}
          | {:message, map()}
          | :noreply

  # ── Public API ────────────────────────────────────────────────────────────

  def start_link(opts \\ []) do
    GenServer.start_link(__MODULE__, {self(), opts})
  end

  @doc "Returns the Unix socket path the fake host listens on."
  def socket_path(host), do: GenServer.call(host, :socket_path)

  @doc "Encodes and sends one JSON-RPC message to the guest."
  def send_message(host, message), do: GenServer.call(host, {:send_raw, frame(message)})

  @doc "Sends raw bytes to the guest. Use it to script partial or joined frames."
  def send_raw(host, bytes), do: GenServer.call(host, {:send_raw, bytes})

  @doc "Returns every byte the guest sent, in order."
  def received_bytes(host), do: GenServer.call(host, :received_bytes)

  @doc "Closes the accepted connection."
  def close(host), do: GenServer.call(host, :close)

  @doc "Frames a message with the LSP `Content-Length` header."
  def frame(message) when is_map(message), do: frame(JSON.encode!(message))

  def frame(body) when is_binary(body) do
    "Content-Length: #{byte_size(body)}\r\n\r\n" <> body
  end

  @doc "Waits for the next request the guest sent and returns it."
  def await_request(timeout \\ 1000) do
    receive do
      {:fake_host, :request, message} -> message
    after
      timeout -> raise "no request from the guest after #{timeout}ms"
    end
  end

  # ── GenServer ─────────────────────────────────────────────────────────────

  @impl true
  def init({owner, opts}) do
    path = Keyword.get_lazy(opts, :socket_path, &temp_socket_path/0)
    File.rm(path)

    {:ok, listen} =
      :gen_tcp.listen(0, [
        :binary,
        {:ifaddr, {:local, path}},
        active: false,
        packet: :raw,
        reuseaddr: true,
        backlog: 1
      ])

    server = self()

    {:ok, _acceptor} =
      Task.start_link(fn ->
        case :gen_tcp.accept(listen) do
          {:ok, socket} ->
            :ok = :gen_tcp.controlling_process(socket, server)
            send(server, {:accepted, socket})

          {:error, _reason} ->
            :ok
        end
      end)

    state = %{
      owner: owner,
      handler: Keyword.get(opts, :handler, fn _ -> :noreply end),
      path: path,
      listen: listen,
      socket: nil,
      buffer: "",
      received: ""
    }

    {:ok, state}
  end

  @impl true
  def handle_call(:socket_path, _from, state), do: {:reply, state.path, state}

  def handle_call(:received_bytes, _from, state), do: {:reply, state.received, state}

  def handle_call({:send_raw, bytes}, _from, state) do
    :ok = :gen_tcp.send(state.socket, bytes)
    {:reply, :ok, state}
  end

  def handle_call(:close, _from, state) do
    if state.socket, do: :gen_tcp.close(state.socket)
    :gen_tcp.close(state.listen)
    File.rm(state.path)
    {:reply, :ok, %{state | socket: nil}}
  end

  @impl true
  def handle_info({:accepted, socket}, state) do
    :ok = :inet.setopts(socket, active: :once)
    {:noreply, %{state | socket: socket}}
  end

  def handle_info({:tcp, socket, data}, state) do
    {frames, buffer} = split_frames(state.buffer <> data, [])
    state = %{state | buffer: buffer, received: state.received <> data}
    state = Enum.reduce(frames, state, &handle_frame/2)
    :ok = :inet.setopts(socket, active: :once)
    {:noreply, state}
  end

  def handle_info({:tcp_closed, _socket}, state) do
    send(state.owner, {:fake_host, :closed})
    {:noreply, %{state | socket: nil}}
  end

  def handle_info(_message, state), do: {:noreply, state}

  @impl true
  def terminate(_reason, state) do
    if state.socket, do: :gen_tcp.close(state.socket)
    :gen_tcp.close(state.listen)
    File.rm(state.path)
    :ok
  end

  # ── Internals ─────────────────────────────────────────────────────────────

  defp handle_frame(body, state) do
    message = JSON.decode!(body)
    send(state.owner, {:fake_host, :request, message})

    case state.handler.(message) do
      {:result, value} ->
        reply(state, %{"jsonrpc" => "2.0", "id" => message["id"], "result" => value})

      {:error, value} ->
        reply(state, %{"jsonrpc" => "2.0", "id" => message["id"], "error" => value})

      {:message, custom} ->
        reply(state, custom)

      :noreply ->
        state
    end
  end

  defp reply(state, message) do
    :ok = :gen_tcp.send(state.socket, frame(message))
    state
  end

  defp split_frames(buffer, acc) do
    case :binary.split(buffer, "\r\n\r\n") do
      [_incomplete] ->
        {Enum.reverse(acc), buffer}

      [headers, rest] ->
        length = content_length(headers)

        if byte_size(rest) >= length do
          <<body::binary-size(length), remaining::binary>> = rest
          split_frames(remaining, [body | acc])
        else
          {Enum.reverse(acc), buffer}
        end
    end
  end

  defp content_length(headers) do
    headers
    |> String.split("\r\n")
    |> Enum.find_value(fn line ->
      case String.split(line, ":", parts: 2) do
        [name, value] ->
          if String.downcase(String.trim(name)) == "content-length" do
            String.to_integer(String.trim(value))
          end

        _ ->
          nil
      end
    end)
  end

  defp temp_socket_path do
    unique = System.unique_integer([:positive])
    base = System.tmp_dir!()
    path = Path.join(base, "ah#{unique}.sock")

    # macOS limits a Unix socket path to 104 bytes.
    if byte_size(path) < 100, do: path, else: Path.join("/tmp", "ah#{unique}.sock")
  end
end

File: tests/Aspire.Hosting.CodeGeneration.Dart.DartTests/test/support/fake_host.dart
// A fake AppHost server for the transport tests.
//
// The fake host listens on a temporary Unix domain socket, accepts one
// connection, and parses LSP framed JSON-RPC messages. Every parsed message
// goes into a queue that [FakeHost.awaitRequest] reads. An optional handler
// scripts the response.

import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'dart:math';
import 'dart:typed_data';

/// Builds the response for one guest message. Return null to send nothing.
typedef FakeHandler =
    Map<String, Object?>? Function(Map<String, Object?> request);

/// Builds a handler that answers each method in [results] with a result.
FakeHandler methodHandler(Map<String, Object?> results) =>
    (Map<String, Object?> request) {
      final Object? method = request['method'];
      if (method is String && results.containsKey(method)) {
        return <String, Object?>{
          'jsonrpc': '2.0',
          'id': request['id'],
          'result': results[method],
        };
      }
      return null;
    };

int _socketCounter = 0;

// `dart test` runs each suite in its own isolate inside one process, so the
// process identifier does not separate two suites. The random prefix does.
final String _socketPrefix = Random().nextInt(0x7fffffff).toRadixString(36);

class FakeHost {
  FakeHost._(this._server, this.socketPath, this._handler);

  /// Starts a fake host on a new temporary socket.
  static Future<FakeHost> start({FakeHandler? handler}) async {
    final String path = _tempSocketPath();
    final File file = File(path);
    if (file.existsSync()) {
      file.deleteSync();
    }

    final ServerSocket server = await ServerSocket.bind(
      InternetAddress(path, type: InternetAddressType.unix),
      0,
    );
    final FakeHost host = FakeHost._(
      server,
      path,
      handler ?? (Map<String, Object?> request) => null,
    );
    host._subscription = server.listen(host._accept);
    return host;
  }

  /// The Unix socket path that the fake host listens on.
  final String socketPath;

  final ServerSocket _server;
  final FakeHandler _handler;
  final Completer<Socket> _accepted = Completer<Socket>();
  final List<int> _buffer = <int>[];
  final List<int> _received = <int>[];
  final List<Map<String, Object?>> _requests = <Map<String, Object?>>[];
  final List<Completer<Map<String, Object?>>> _waiters =
      <Completer<Map<String, Object?>>>[];

  StreamSubscription<Socket>? _subscription;
  StreamSubscription<Uint8List>? _clientSubscription;
  Socket? _client;

  /// Every byte that the guest sent, in order.
  List<int> get receivedBytes => List<int>.unmodifiable(_received);

  /// Frames a message with the LSP `Content-Length` header.
  static List<int> frame(Object message) {
    final List<int> body = utf8.encode(
      message is String ? message : jsonEncode(message),
    );
    return <int>[
      ...utf8.encode('Content-Length: ${body.length}\r\n\r\n'),
      ...body,
    ];
  }

  /// Encodes and sends one JSON-RPC message to the guest.
  Future<void> sendMessage(Map<String, Object?> message) =>
      sendRaw(frame(message));

  /// Sends raw bytes to the guest. Use it to script partial or joined frames.
  ///
  /// The guest can finish `connect` before the listener delivers the accepted
  /// socket. The send waits for that socket instead of failing.
  Future<void> sendRaw(List<int> bytes) async {
    // `close` destroys the accepted socket. This local holds the same socket.
    // ignore: close_sinks
    final Socket socket = await _ensureSocket();
    socket.add(bytes);
    await socket.flush();
  }

  /// Waits for the next message that the guest sent and returns it.
  Future<Map<String, Object?>> awaitRequest({
    Duration timeout = const Duration(seconds: 2),
  }) async {
    if (_requests.isNotEmpty) {
      return _requests.removeAt(0);
    }
    final Completer<Map<String, Object?>> waiter =
        Completer<Map<String, Object?>>();
    _waiters.add(waiter);
    return waiter.future.timeout(
      timeout,
      onTimeout: () {
        _waiters.remove(waiter);
        throw StateError(
          'no message from the guest after ${timeout.inMilliseconds}ms',
        );
      },
    );
  }

  /// Closes the accepted connection and the listening socket.
  Future<void> close() async {
    _client?.destroy();
    _client = null;
    await _clientSubscription?.cancel();
    _clientSubscription = null;
    await _subscription?.cancel();
    _subscription = null;
    await _server.close();
    final File file = File(socketPath);
    if (file.existsSync()) {
      file.deleteSync();
    }
  }

  Future<Socket> _ensureSocket() {
    final Socket? client = _client;
    if (client != null) {
      return Future<Socket>.value(client);
    }
    return _accepted.future.timeout(
      const Duration(seconds: 2),
      onTimeout: () =>
          throw StateError('FakeHost: no client connected within 2s'),
    );
  }

  void _accept(Socket socket) {
    _client = socket;
    _clientSubscription = socket.listen(
      _onData,
      cancelOnError: false,
      onError: (Object _) {},
    );
    if (!_accepted.isCompleted) {
      _accepted.complete(socket);
    }
  }

  void _onData(Uint8List chunk) {
    _received.addAll(chunk);
    _buffer.addAll(chunk);
    while (_drainFrame()) {
      // Every complete frame in the buffer runs before the next read.
    }
  }

  bool _drainFrame() {
    final int separator = _indexOfSeparator();
    if (separator < 0) {
      return false;
    }

    final String headers = latin1.decode(_buffer.sublist(0, separator));
    final int? length = _contentLength(headers);
    if (length == null) {
      _buffer.removeRange(0, separator + 4);
      return true;
    }

    final int start = separator + 4;
    if (_buffer.length - start < length) {
      return false;
    }

    final String body = utf8.decode(_buffer.sublist(start, start + length));
    _buffer.removeRange(0, start + length);
    _handleFrame(body);
    return true;
  }

  int _indexOfSeparator() {
    for (int index = 0; index + 4 <= _buffer.length; index++) {
      if (_buffer[index] == 13 &&
          _buffer[index + 1] == 10 &&
          _buffer[index + 2] == 13 &&
          _buffer[index + 3] == 10) {
        return index;
      }
    }
    return -1;
  }

  static int? _contentLength(String headers) {
    for (final String line in headers.split('\r\n')) {
      final int colon = line.indexOf(':');
      if (colon >= 0 &&
          line.substring(0, colon).trim().toLowerCase() == 'content-length') {
        return int.tryParse(line.substring(colon + 1).trim());
      }
    }
    return null;
  }

  void _handleFrame(String body) {
    final Object? decoded = jsonDecode(body);
    if (decoded is! Map<String, Object?>) {
      return;
    }

    if (_waiters.isNotEmpty) {
      _waiters.removeAt(0).complete(decoded);
    } else {
      _requests.add(decoded);
    }

    final Map<String, Object?>? response = _handler(decoded);
    if (response != null) {
      unawaited(sendRaw(frame(response)));
    }
  }

  static String _tempSocketPath() {
    // macOS limits a Unix socket path to 104 bytes. Keep it short.
    return '/tmp/adt${_socketPrefix}_${++_socketCounter}.sock';
  }
}
